File: 2. Structural Design Patterns/Decorator/Java/DecoratorTemplate.java
// Decorator Pattern – Java Template

interface Component {
    void operation();
}

class ConcreteComponent implements Component {
    public void operation() {
        /* base work */
    }
}

abstract class BaseDecorator implements Component {
    protected final Component wrap;

    BaseDecorator(Component w) {
        this.wrap = w;
    }

    public void operation() {
        wrap.operation();
    }
}

class ConcreteDecoratorA extends BaseDecorator {
    ConcreteDecoratorA(Component w) {
        super(w);
    }

    public void operation() {
        /* extra A */
        super.operation();
    }
}

class ConcreteDecoratorB extends BaseDecorator {
    ConcreteDecoratorB(Component w) {
        super(w);
    }

    public void operation() {
        /* extra B */
        super.operation();
    }
}

public class DecoratorTemplate {
    public static void main(String[] args) {
        Component c = new ConcreteDecoratorB(
                new ConcreteDecoratorA(new ConcreteComponent())
        );
        c.operation();
    }
}

File: 2. Structural Design Patterns/Decorator/Kotlin/DecoratorTemplate.kt
// Decorator Pattern – Kotlin Template

interface Component { fun operation() }
class ConcreteComponent : Component { override fun operation() { /* base work */ } }
open class BaseDecorator(protected val wrap: Component) : Component { override fun operation() = wrap.operation() }
class ConcreteDecoratorA(w: Component) : BaseDecorator(w) { override fun operation(){ /* extra A */ super.operation() } }
class ConcreteDecoratorB(w: Component) : BaseDecorator(w) { override fun operation(){ /* extra B */ super.operation() } }

fun main(){ val c: Component = ConcreteDecoratorB(ConcreteDecoratorA(ConcreteComponent())); c.operation() }
